patterns/self_encapsulation.md

```diff
@@ -0,0 +1,655 @@
+
+## The individual patterns are:
+
+* Behavior-Defined Class
+* Abstract State
+* Message Layers
+* Deferred State Variables
+* Encapsulated Concrete State
+* Lazy Initialization
+* Default Value Method
+* The Pattern Language
+
+## Behavior-Defined Class
+_**Also Known As: Define Classes by Behavior, Not State**_
+
+### Context:
+
+You are creating a new class. You want your implementation to maximize the Reusability of your class and its subclasses as they evolve.
+
+### Problem:
+
+Typically, we want subclasses to inherit *behavior*. How do we define classes so that behavior is reliable?
+
+### Constraints:
+
+In a collaborating system of objects, the essence of an object is its responsibilities. Other objects rely on it to fulfill these responsibilities. How it fulfills them is relatively unimportant. Even less important is an object's data structure (that is, the number of slots*, the slots' names, and the type of object that is expected to live in those slots). First and foremost, the essence of an object is its responsibilities.
+
+An object fulfills its responsibilities by passing messages*. Programmers usually find it a lot less time-consuming to define message names (protocols) than to define the functional details of how an object responds to them. (Note: I refer here to the time necessary to define these names for the rest of the system. Don't confuse that with the time needed to determine high-quality message names, which should always be done with care.)
+
+Additionally, the object's data structure often plays an important role in how an object responds to the messages it receives. An object cannot do much without eventually manipulating some sort of data. Implementing an object's data structure forces the programmer to make certain trade-offs between memory space and speed. Considering those trade-offs can distract a programmer from a more serious implication of their decisions: **every data structure implemented in a base class affects its subclass.**
+
+Subclasses automatically inherit the data structure of their base class. So, any decisions about data structure are virtually irreversible by subclasses without overriding every base class method that refers to the data structure. Therefore, each time you add detail to the data structure, some implementation options are lost and others imposed on the class and its subclasses. And, the amount of time it takes to refactor anything involving the data structure of a base class is quickly multiplied by the number subclasses affected by the change.
+
+* Slots could refer to storage slots of any kind (memory, database, etc.)
```

Comment:
Perhaps an improvement on this sentence:
```
When we say _slot_ we mean some slot in memory or a database allocated to hold a specific piece of data.  It could be a named property or instance_variable, a cell in a database, an indexed variable in an array, a key/value pair, etc.
```